File: bank_campaign_classification.py
# Bank Campaign Customer Classification
# Authors: Elbekova Aidai, Bethelhem Samson Gebreegziabhier, Joshua Adu
# Supervisors: Prof. Dr. André Hanelt, Steven Görlich, M. Sc.

import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV, KFold, cross_val_score
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score
from imblearn.over_sampling import RandomOverSampler

# Load the dataset
dataset = pd.read_csv("bank-full.csv", sep=";")

# Data Preprocessing
# Handling missing values
dataset.fillna(method='ffill', inplace=True)

# Encoding Categorical Data
categorical_features = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']
numeric_features = dataset.select_dtypes(include=['int64', 'float64']).columns.to_list()
numeric_features.remove('y')

# One-hot encoding
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_features),
        ('cat', OneHotEncoder(), categorical_features)])

# Splitting the Dataset
X = dataset.drop('y', axis=1)
y = dataset['y'].apply(lambda x: 1 if x == 'yes' else 0)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

# Random Oversampling
ros = RandomOverSampler(random_state=123)
X_train_res, y_train_res = ros.fit_resample(X_train, y_train)

# Model Description
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Random Forest": RandomForestClassifier(),
    "Support Vector Machine": SVC(probability=True),
    "Neural Network": MLPClassifier(max_iter=1000),
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Ada Boost": AdaBoostClassifier(),
    "Naive Bayes": GaussianNB(),
    "Decision Trees": DecisionTreeClassifier()
}

# Train and evaluate each model
for name, model in models.items():
    pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                               ('classifier', model)])
    
    # Hyperparameter tuning (example with Random Forest)
    if name == "Random Forest":
        param_grid = {
            'classifier__n_estimators': [10, 50, 100],
            'classifier__max_features': ['auto', 'sqrt', 'log2']
        }
        grid_search = GridSearchCV(estimator=pipeline, param_grid=param_grid, cv=5, scoring='roc_auc')
        grid_search.fit(X_train_res, y_train_res)
        best_model = grid_search.best_estimator_
        y_pred = best_model.predict(X_test)
        print(f"\n{name} - Best Params: {grid_search.best_params_}")
    else:
        pipeline.fit(X_train_res, y_train_res)
        y_pred = pipeline.predict(X_test)
    
    # Evaluation
    print(f"\n{name} - Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
    print(f"{name} - Classification Report:\n", classification_report(y_test, y_pred))
    print(f"{name} - ROC AUC Score: {roc_auc_score(y_test, y_pred)}")

# Save the best model for later use (example with Random Forest)
import joblib
joblib.dump(best_model, "model_rf_opt.pkl")

File: bank_campaign_classification.R
# Bank Campaign Customer Classification
# Authors: Elbekova Aidai, Bethelhem Samson Gebreegziabhier, Joshua Adu
# Supervisors: Prof. Dr. André Hanelt, Steven Görlich, M. Sc.

# Load necessary libraries
library(tidyverse)
library(caret)
library(e1071)
library(randomForest)
library(nnet)

# Load the dataset
dataset <- read.csv("bank-full.csv", sep=";")

# Data Preprocessing

# Finding Missing Values
missing_values <- sapply(dataset, function(x) sum(is.na(x)))
print(missing_values)

# Encoding Categorical Data
dataset$job <- as.factor(dataset$job)
dataset$marital <- as.factor(dataset$marital)
dataset$education <- as.factor(dataset$education)
dataset$default <- as.factor(dataset$default)
dataset$housing <- as.factor(dataset$housing)
dataset$loan <- as.factor(dataset$loan)
dataset$contact <- as.factor(dataset$contact)
dataset$month <- factor(dataset$month, levels = c("jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"))
dataset$day_of_week <- factor(dataset$day_of_week, levels = c("mon", "tue", "wed", "thu", "fri"))
dataset$poutcome <- as.factor(dataset$poutcome)
dataset$y <- as.factor(dataset$y)

# One-hot encoding
dataset <- dummyVars(" ~ .", data = dataset)
dataset <- data.frame(predict(dataset, newdata = dataset))

# Splitting the Dataset
set.seed(123)
trainIndex <- createDataPartition(dataset$y, p = .8, 
                                  list = FALSE, 
                                  times = 1)
train_data <- dataset[ trainIndex,]
test_data  <- dataset[-trainIndex,]

# Feature Scaling
preProcValues <- preProcess(train_data, method = c("center", "scale"))
train_data <- predict(preProcValues, train_data)
test_data <- predict(preProcValues, test_data)

# Random Oversampling
train_data <- upSample(x = train_data, y = train_data$y)

# Model Description

# Logistic Regression
model_lr <- train(y ~ ., data = train_data, method = "glm", family = "binomial")
predictions_lr <- predict(model_lr, test_data)
confusionMatrix(predictions_lr, test_data$y)

# Random Forest
model_rf <- train(y ~ ., data = train_data, method = "rf")
predictions_rf <- predict(model_rf, test_data)
confusionMatrix(predictions_rf, test_data$y)

# Support Vector Machine
model_svm <- train(y ~ ., data = train_data, method = "svmRadial")
predictions_svm <- predict(model_svm, test_data)
confusionMatrix(predictions_svm, test_data$y)

# Neural Network
model_nn <- train(y ~ ., data = train_data, method = "nnet", linout = TRUE)
predictions_nn <- predict(model_nn, test_data)
confusionMatrix(predictions_nn, test_data$y)

# K-Nearest Neighbors
model_knn <- train(y ~ ., data = train_data, method = "knn")
predictions_knn <- predict(model_knn, test_data)
confusionMatrix(predictions_knn, test_data$y)

# Ada Boost
model_ada <- train(y ~ ., data = train_data, method = "adaboost")
predictions_ada <- predict(model_ada, test_data)
confusionMatrix(predictions_ada, test_data$y)

# Naive Bayes
model_nb <- train(y ~ ., data = train_data, method = "nb")
predictions_nb <- predict(model_nb, test_data)
confusionMatrix(predictions_nb, test_data$y)

# Decision Trees
model_dt <- train(y ~ ., data = train_data, method = "rpart")
predictions_dt <- predict(model_dt, test_data)
confusionMatrix(predictions_dt, test_data$y)

# Data Optimization
# Hyperparameter tuning and k-fold validation (example with Random Forest)

control <- trainControl(method="cv", number=10)
tunegrid <- expand.grid(.mtry=c(1:15))
model_rf_opt <- train(y ~ ., data=train_data, method="rf", trControl=control, tuneGrid=tunegrid)
predictions_rf_opt <- predict(model_rf_opt, test_data)
confusionMatrix(predictions_rf_opt, test_data$y)

# Print the optimized Random Forest model details
print(model_rf_opt)

# Save the models for later use
saveRDS(model_rf, "model_rf.rds")
saveRDS(model_rf_opt, "model_rf_opt.rds")
